File: app/src/main/java/ds/vkplus/ui/activity/MainActivity.java
package ds.vkplus.ui.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import ds.vkplus.R;
import ds.vkplus.model.ApiResponse;
import ds.vkplus.network.RestService;
import ds.vkplus.utils.L;


public class MainActivity extends AppCompatActivity {


	@Override
	protected void onCreate(Bundle savedInstanceState) {
		supportRequestWindowFeature(Window.FEATURE_INDETERMINATE_PROGRESS);
		supportRequestWindowFeature(Window.FEATURE_ACTION_BAR_OVERLAY);
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
		//setTitle(R.string.news);

		ActionBar ab = getSupportActionBar();
		//ab.setHideOnContentScrollEnabled(true);
		ab.setIcon(R.drawable.logo);
		View v=getActionBarView();

		if (savedInstanceState == null) {
			showFragment();
			//RestService.get().login(this).subscribe(token -> showFragment(), e -> T.show(this, "Failed to login"));
		}

		RestService.get()
		           .getGroups()
		           .subscribe(groups -> {
			           L.v("groups fetched successfully");
		           }, Throwable::printStackTrace);
	}

	public View getActionBarView() {
		Window window = getWindow();
		View v = window.getDecorView();
		int resId = getResources().getIdentifier("action_bar_container", "id", "android");
		return v.findViewById(resId);
	}


	private void showFragment() {
		getSupportFragmentManager().beginTransaction()
		                           .replace(R.id.container, getFragment())
		                           .commit();
	}


	private Fragment getFragment() {
		return new NewsFragment();
	}


	@Override
	public boolean onCreateOptionsMenu(final Menu menu) {
		getMenuInflater().inflate(R.menu.main, menu);
		return super.onCreateOptionsMenu(menu);
	}


	@Override
	public boolean onOptionsItemSelected(final MenuItem item) {
		switch (item.getItemId()) {
			case R.id.debug_rest:
				RestService.get()
				           .dummyRequestWithError(new ApiResponse.Error(5, "fake error!"))
				           .subscribe(
						           L::v,
						           e -> {
							           L.e("last mile error");
							           e.printStackTrace();
						           },
						           () -> L.v("completed"));

				break;

			case R.id.debug_layout:
				startActivity(new Intent(this,TempActivity.class));
				break;

		}
		return super.onOptionsItemSelected(item);
	}
}

File: app/src/main/java/ds/vkplus/ui/activity/MainActivity.kt
package ds.vkplus.ui.activity

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.Window
import com.trello.rxlifecycle.components.support.RxAppCompatActivity
import ds.vkplus.R
import ds.vkplus.model.ApiResponse
import ds.vkplus.network.RestService
import ds.vkplus.utils.L
import ds.vkplus.utils.Utils
import rx.android.schedulers.AndroidSchedulers
import rx.lang.kotlin.observable
import rx.schedulers.Schedulers


class MainActivity : RxAppCompatActivity() {

	var newTransitionName: String? = null
	
	override fun onCreate(savedInstanceState: Bundle?) {
		supportRequestWindowFeature(Window.FEATURE_ACTION_BAR_OVERLAY)
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_main)

		supportActionBar?.setIcon(R.drawable.logo)
		supportActionBar?.setDisplayShowTitleEnabled(false)

		if (savedInstanceState == null) {
			showFragment()
		}
		
		//RestService.instance.getGroups().subscribe({ L.v("groups: fetched successfully") }, { it.printStackTrace() })
	}
	
	fun getActionBarView(): View {
		val window = window
		val v = window.decorView
		val resId = resources.getIdentifier("action_bar_container", "id", "android")
		return v.findViewById(resId)
	}
	
	
	private fun showFragment() {
		observable<ProgressDialog> {
			val progressDialog = ProgressDialog.show(this@MainActivity, null, "Loading...")
			progressDialog.setCancelable(true)
			it.onNext(progressDialog)
		}
			.observeOn(Schedulers.computation())
			.doOnNext {
				if (Utils.isMainThread()) throw IllegalThreadStateException()
				RestService.instance.getGroups().toBlocking().first()
			}
			.doOnError { it.printStackTrace() }
			.observeOn(AndroidSchedulers.mainThread())
			.compose<ProgressDialog> (bindToLifecycle())
			.subscribe ({
				it.dismiss()
				supportFragmentManager.beginTransaction().replace(R.id.container, getFragment(), "news").commit()
			},{
				L.e("failed to attach fragment")
				it.printStackTrace()
			})
	}
	
	
	private fun getFragment(): Fragment {
		return NewsFragment()
	}
	
	
	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		menuInflater.inflate(R.menu.main, menu)
		return super.onCreateOptionsMenu(menu)
	}
	
	
	override fun onOptionsItemSelected(item: MenuItem): Boolean {
		when (item.itemId) {
			R.id.debug_rest -> RestService.instance.dummyRequestWithError(ApiResponse.Error(5, "fake error!"))
				.subscribe({
					L.v(it)
				}, {
					L.e("last mile error")
					it.printStackTrace()
				}, {
					L.v("completed")
				})
			
			//R.id.debug_layout -> startActivity(Intent(this, TempActivity::class.java))
		}
		return super.onOptionsItemSelected(item)
	}

	override fun onActivityReenter(requestCode: Int, data: Intent) {
		super.onActivityReenter(requestCode, data)
		val id = data.extras.getString("id")
		newTransitionName = id
	}


}
